### DISCUSSION TITLE
Weird initialisation behaviour for a parameter record

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#9320] [Q&A] [answered]
I recently found some puzzling behaviour when initialising a thermodynamic reference state (i.e., a parameter record).
When simulating the following MWE in OM 1.19.2, `state_ref1.T` gets the value 500 K (226.85°C), not 333 K as expected from the `initial equation`. I don't understand why this happens or where 500 K comes from. 
If I set the parameter via declaration equation (`state_ref2`), results are expected.

Questions:
1. Before opening an issue about this, I thought I'd ask if I'm overlooking some specification detail that I'm unaware of.
2. So far I thought that these two ways of setting parameters (`final parameter` with declaration equation vs. plain `parameter`+`initial equation`) are roughly equivalent/interchangeable for fixed/not user-modifiable parameters. Am I wrong?
3. Probably connected with remark 2, in the Plotting perspective, `T` and `p` values of `state_ref1` can be modified and are preserved on resimulation, which afaict overrides/disables the initial equation! I would have expected the initial equation to still be active on resimulation, and reset `state_ref1` to the prescribed values. Why does it not?
4. I tried to set `fixed=false` for the parameter state_ref1, but only got "Modified element fixed not found in class ThermodynamicState." (of course, I guess). The specification in 8.6 says 

> "All variables declared as parameter having fixed = false are treated as unknowns during the initialization phase, i.e. there must be additional equations for them – and the start-value can be used as a guess-value during initialization.

and a variable is (Spec 4.4.) "an instance (object) generated by a component declaration." so I thought it should be possible to set `fixed` false. How does that work, or does that not make sense/is not necessary in case of records?

```
model OM_state_init

  package Medium = Modelica.Media.Air.DryAirNasa;
  parameter Medium.AbsolutePressure p_ref=10e5;
  parameter Medium.Temperature T_ref=333;
  parameter Medium.MassFraction X_ref[:] = Medium.X_default;
  
  // AFAIK, both ways of setting the parameter should be equivalent
  // The resulting pressure in the states is correct, but the temperature 
  // in state_ref1 is 500 K, not 333!
  parameter Medium.ThermodynamicState state_ref1;
  final parameter Medium.ThermodynamicState state_ref2 = Medium.setState_pTX(p_ref, T_ref, X_ref);
initial equation
  state_ref1 = Medium.setState_pTX(p_ref, T_ref, X_ref);
end OM_state_init;
```

**Comment by AnHeuermann** [comment, marked as answer]
Looks like a bug to me. And one I would like to fix because I had similar issues in other models and there the initialization is still broken.

**Reply by AnHeuermann** [reply]
Created #9321.

**Reply by bilderbuchi** [reply]
Thanks!

**Comment by AnHeuermann** [comment]
In my opinion this is still a bug, but in the sense that the user should get an error/warning that the initial system is overspezified because final parameters have `fixed=true`, so the default value (500) is used.

In any case, here is something both OpenModelica and Dymola accept:

```Modelica
model OM_state_init
  package Medium = Modelica.Media.Air.DryAirNasa;
  parameter Medium.AbsolutePressure p_ref=10e5;
  parameter Medium.Temperature T_ref=333;
  parameter Medium.MassFraction X_ref[:] = Medium.X_default;
  
  parameter Medium.ThermodynamicState state_ref1(T(fixed=false), p(fixed=false));
  final parameter Medium.ThermodynamicState state_ref2 = Medium.setState_pTX(p_ref, T_ref, X_ref);
initial equation
  state_ref1 = Medium.setState_pTX(p_ref, T_ref, X_ref);
end OM_state_init;
```

**Reply by bilderbuchi** [reply]
>  because final parameters have fixed=true,

AFAIK the "final" is not necessary here, parameters by default have `fixed=true`?
Anyway, nice technique (potentially cumbersome, though) for getting a fixed=false record -- setting all members individually. I didn't think of that.

**Comment by casella** [comment]
Some additional comments. I agree the original model is buggy: a parameter without a binding equation still has fixed = true, so it lacks a binding equations, which cannot be replaced by an initial equation.

Setting all members variables to fixed = false works, but only if you don't use a replaceable medium, otherwise you can't know what are the actual member variables after redeclare.

If you really want to use initial equations, you can declare state_ref1 as a discrete variable, instead of a parameter. In this case, it is ok to use an initial equation to set a value to it.

Adding "final" doesn't change the equation count. It only tells the compiler that the binding cannot be undone at a later time (runtime, or upon instantiation/redeclaration), which allows the compiler to carry on further optimizations by exploiting the final binding.

**Reply by bilderbuchi** [reply]
> Adding "final" doesn't change the equation count. It only tells the compiler that the binding cannot be undone at a later time (runtime, or upon instantiation/redeclaration), which allows the compiler to carry on further optimizations by exploiting the final binding.

The intent behind the final was to prohibit reassigning the parameter a different value via modification, which would not make sense (see my original point 2. above). I (think I) understand now that to have equivalence between `final parameter` and `parameter` + `initial equation`, the latter would need `fixed=false` (and does not work in all cases).

**Reply by casella** [reply]
Exactly
